src/fcntl.rs

```diff
@@ -692,6 +692,9 @@ pub enum FcntlArg<'a> {
     /// Return the full path without firmlinks of the fd.
     #[cfg(apple_targets)]
     F_GETPATH_NOFIRMLINK(&'a mut PathBuf),
+    /// Pre-allocate storage with different policies on fd.
+    #[cfg(apple_targets)]
+    F_PREALLOCATE(&'a libc::fstore_t),
```

Comment:
The associated buffer has an `OUT` field: 

```c
typedef struct fstore {
	    u_int32_t fst_flags;      /* IN: flags word */
	    int       fst_posmode;    /* IN: indicates offset field */
	    off_t     fst_offset;     /* IN: start of the region */
	    off_t     fst_length;     /* IN: size of the region */
	    off_t     fst_bytesalloc; /* OUT: number of bytes allocated */
} fstore_t;
```

i.e., the buffer can be written, if we use an immutable reference here, then UB could happen


Ref: https://keith.github.io/xcode-man-pages/fcntl.2.html#F_PREALLOCATE

Reply:
I just tried implementing this on my mac, I think we should create a helper type for those `IN` fields, and return the `OUT` field through `fcntl()`.

For that `fst_falgs` field, we would create a bitflags struct, for `fst_posmode`, a Rust enum would suffice.


And [F_ALLOCATEPERSIST](https://keith.github.io/xcode-man-pages/fcntl.2.html#F_ALLOCATEPERSIST) is not available in libc, we should add it first.

Reply:
libc PR filed: https://github.com/rust-lang/libc/pull/3713